rfcs/2021-12-20-10517-llvm-backend-for-vrl.md

```diff
@@ -0,0 +1,560 @@
+
+Producing LLVM bitcode for Rust's `std` library is guarded behind the [`-Z build-std`](https://doc.rust-lang.org/cargo/reference/unstable.html#build-std) flag and only available on the nightly compiler toolchain. We need `std` to fully link our precompiled LLVM bitcode. Further investigation is needed here if this means that we need to upgrade our entire compiler toolchain to nightly or if it's possible to build a binary-compatible `std` according to our `rust-toolchain.toml` file.
+
+Statically linking LLVM to the Vector binary adds roughly 9MB, additionally to precompiled bitcode that needs to be included with the binary. If this is a concern, we can consider shipping binaries with the LLVM feature disabled.
+
+While LLVM is a highly used framework within the industry, working with it requires rather specialized knowledge about compiler construction. However, there exists plenty of publicly accessible material for code generation using LLVM, some of which I linked to [above](#introduction-to-llvm). In addition, there should be great focus to document this part of the code base extraordinarily well.
+
+## Prior Art
+
+_TODO._
+
+## Alternatives
+
+### Compile to Rust
+
+Using Rust as a compilation target would require us to
+
+- ship a Rust compiler and its libraries
+- ship Vector source code and its dependent crates
+
+which would be hundreds of MB and therefore infeasible.
+
+### Compile to C
+
+Using C as a compilation target would require us to
+
+- ship a C compiler
+
+while
+
+- not having any better safety guarantees
+- not being able to inline functions and therefore miss optimization potential
+
+and therefore not provide any significant benefits over using LLVM directly.
+
+### Compile to WebAssembly
+
+Using WebAssembly as a compilation target would require us to
+
+- ship a Wasm runtime
+- copy data in and out of WebAssembly or use `mmap`ing techniques which would constrain in which memory regions event data must reside
+- likely slower execution speed because of higher abstraction level and semantics that allow untrusted code to execute safely
+- precompile WebAssembly bitcode or miss optimization potential
+
+additionally to these drawbacks, we just dropped support for the WebAssembly transform.
+
+### Compile to Bitcode
+
+As mentioned in the context, we are currently moving forward with a VM for VRL. Compared to the current execution model and an LLVM-based approach, the VM provides a middle ground for execution speed, memory safety and sophistication.
+
+Weighing the benefits depends on the real world performance of both approaches.
+
+## Plan Of Attack
+
+Incremental steps to execute this change. These will be converted to issues after the RFC is approved:
+
+- Submit a PR with spike-level code _roughly_ demonstrating the change: [#10442](https://github.com/vectordotdev/vector/pull/10442).
```

Comment:
My only ask for this plan of attack would be to emphasize getting soak numbers as rapidly as possible, mostly because that helps us figure out how much of a boon this is over the byte code VM and/or where bottlenecks appear in surrounding Vector. I think that's implicit here from chatting, but figured it'd be worth saying out loud.

Reply:
Yep, this is definitely my first priority. Just had to spend the time to get the implementation to a point where it can run programs meaningful enough to compare

Reply:
Added a specific action item in b2c888b0df7fce20ba531ce08431b05ff86e8900.